### PR TITLE
DS3: Version 3.0.0

### PR DESCRIPTION
**Pull request by nex3**
This is a major rewrite of the _Dark Souls III_ apworld, as discussed initially in Marechal-L/Dark-Souls-III-Archipelago-client#17 and in the Discord channel. It integrates with [a branch] of @thefifthmatt's static DS3 randomizer to enable a broad swath of new capabilities and improved player experience.

[a branch]: https://github.com/nex3/SoulsRandomizers/tree/archipelago-server

## What is this fixing or adding?

* Support for randomizing all item locations, not just unique items, without needing progressive pickup lists.

* Support for randomizing items in shops, starting loadouts, Path of the Dragon, and more.

* Built-in integration with the enemy randomizer, including consistent seeding for races and configuration via the web interface and the standard Archipelago YAML file.

* Support for the latest patch for Dark Souls III, 1.15.2.

* Optional smooth distribution for upgrade items, upgraded weapons, and soul items so you're more likely to see weaker items earlier and more powerful items later.

* More detailed location names that indicate where a location is, not just what it replaces.

* Other players' item names are visible in DS3.

* If you pick up items while offline, they'll still send once you reconnect.

## How was this tested?

I've run several multiworlds with this version, as have a number of beta testers from the DS3 channel on the Discord. @Exempt-Medic has run many hundreds of randomized generations with it to flush out edge-case bugs in generation logic.

To test locally, take use [the latest beta release](https://github.com/nex3/Dark-Souls-III-Archipelago-client/releases) as the client and follow the instructions it provides.

### CODE REVIEW

**Comment by nex3** [comment]
Re: licensing, @gracenotes (who maintains the static randomizer) is on board to upstream my changes, so there should be no issue there in the long term. (I believe, although Matt can correct me on this, that the reasoning behind it not being freely licensed is that it uses/modifies some existing libraries that never got real licenses and whose original authors are now incommunicado.)

I'm happy to explain any changes, but I'm not sure which you mean specifically. I think most of the changes here are pretty clearly driven by the new features I listed above.

**Comment by Exempt-Medic** [comment]
> Additionally, you should document the reasoning behind the couple of core changes in this PR, as their relevance is unclear.

The `Generate.py` changes are already in [PR 2403](https://github.com/ArchipelagoMW/Archipelago/pull/2403).

I do think the `flatten` changes could use some explanation though. They are also probably meant to be used in [Items.py](https://github.com/ArchipelagoMW/Archipelago/pull/3128/files#diff-918dd4987751aff5238ae5bc5dbdbf7998e25e8bd45b4d3a84c6d30b6b37a0d2R269) though they are not

**Comment by nex3** [comment]
Oh, the changes outside of DS3! I understand now. Yes, as @Exempt-Medic says, the `Utils.py` changes are just pulling in https://github.com/ArchipelagoMW/Archipelago/pull/2403. The `flatten()` change is taking a utility function that was used by SM and pulling it into `Utils.py` so that other worlds can use it. (The reason DS3 doesn't currently is so that it could be bundled into a standalone apworld, but it should start using it before this lands.)

File: worlds/sm/variaRandomizer/logic/smbool.py
Comment (lordlou) [inline worlds/sm/variaRandomizer/logic/smbool.py:1]: I have to kindly oppose to this change because all code under /variaRandomizer/ comes from VARIA and each additional change will make future update harder. Also, I dont like adding a dependency to a core module unless essential. Feel free to copy that function though.

**Review by nicholassaylor — CHANGES_REQUESTED** [review]
## Nature of Review
This review mainly looks at documentation and options, focusing on readability, simplicity, and conciseness. Because of this, my primary focus was on all of the markdown files and Options references in both `Options.py` and `__init__.py`. 

This review does not cover changes to core or other worlds, non-option changes to `__init__.py`, the auto-generation for documentation added to DS3, or any changes to the Bosses, Items, or Locations files. 

Tests were not run locally or reviewed due to the nature of this review.

### The Good
The updated documentation is amazing, probably the most comprehensive of any of the games in main currently. This effort to include all randomizer-relevant information causes the various pages to feel a bit bloated, but it does not get in the way of the user experience.

The options added are very well explained and cover a wide range of playstyles for DS3. Looking through the new options, I did not see anything of particular concern as far as implementation.

### The Bad
The decision to keep deprecated options was puzzling to me since the integration effectively duplicates the options between the various "Randomize ___" options and excluded locations. 

I recommend trashing the old options entirely since they end up cluttering the options page and template YAML. Since users have to update their client and game anyway, I see no reason to try to hold onto the old options. 

I am more than willing to hear the rationale on this decision, but with how it is presented in the AP code, I don't understand its purpose.

### The Ugly
While the documentation is very comprehensive, its density can lead to it feeling bloated despite containing all relevant information. In particular, the locations page is extremely information dense and is very long due to the inclusion of the location table at the bottom. See my comments on those files for more context.

Additionally, there are markdown files included that are only accessible through links on the game page rather than being available like the options and set-up guides. Part of this is due to how the WebHost currently handles pages for each game, however I see no reason to try and change that for the sake of more documentation being supported.

### Key Points
* Deprecated options should be removed
* Some concerns about expansive documentation, detailed in review comments
* Minor spelling and clarification fixes in documentation
* Other small nitpicks (can safely be closed without explanation)

File: worlds/dark_souls_3/__init__.py
Comment (qwint) [inline worlds/dark_souls_3/__init__.py:54]: I know this wasn't changed in the PR but I don't think this is important for DS as the regions are not shuffled so the paths should not need to be present in the spoiler log

**Review by qwint — CHANGES_REQUESTED** [review]
Wanted to note that while the default exclude_locations populates correctly on Launcher's generate templates, the webhost export settings on defaults does not include it (on options page and weighted options page). I'm not sure how to make it function without that being part of the webhost rework though, but could be documented by the other text describing the defaults.

skimmed the data files and docs and the structure looked reasonable but didn't interrogate further
also speedread through docs for any typos/inconsistencies, and answered all of my ignorant questions quickly, with no other notes

I don't like the Generate and Utils changes being here and not in another, dependent, Core PR, and from some other conversations after i started this review it looks like they either are or should be removed; which i agree with

tests look fine, ran them on 3.8/3.11 no issue

Only briefly looked over the actual access rules applied (as i have little game knowledge) but did review the helper functions to confirm they are being sensible so anyone with more game knowledge could review the set rules code easily.

did run through some of the benchmarks in test/benchmark and load_worlds was a respectable time. But running benchmark/locations.py, some of the location benchmarks are > 3 seconds in all_state. they seem to be mostly quests that require multiple different can_reach(location) for other locations, which implies that they could be simplified by adding those access rules onto the relevant location's rules and setting some can_reach(region) instead. But i would want more testing to be sure that would actually help.

didn't fully grock the creations of the ordered lists in the item smoothing section, but assuming they are created correct the rest looks good

Also of note: didn't check anything like unused imports etc. or play through the game at all

File: worlds/dark_souls_3/docs/locations_en.md
Comment (nex3) [reply worlds/dark_souls_3/docs/locations_en.md:755]: I could move it into yet another different page, but then it's another pager players have to navigate to. I think it makes more sense to just have _all_ the density here. Since this is at the bottom of the page anyway, I think the actual cost of including it is pretty low despite how dense it is.

**Comment by nex3** [comment]
@PoryGone That's fair, although my other small separate-pr change #2403 has been languishing for six months so I'm a bit nervous about splitting stuff out. I'm happy to pull that out separately before this lands, though.

@lordlou Oh, I didn't realize that came from a separate source. I've reverted it to use its own flatten function.

@nicholassaylor I'd be thrilled to remove the deprecated options. I kept them because I wanted this to be as minimally disruptive as possible (which, to be fair, isn't _that_ possible). But I would personally prefer to have them gone, so if that's the consensus here as well I'll remove them happily.

@qwint

> Wanted to note that while the default exclude_locations populates correctly on Launcher's generate templates, the webhost export settings on defaults does not include it (on options page and weighted options page). I'm not sure how to make it function without that being part of the webhost rework though, but could be documented by the other text describing the defaults.

I had a PR that made this work, but I think it was superseded by the current ongoing webhost work. If this looks likely to land before the rework, I'll add a note.

File: worlds/dark_souls_3/docs/locations_en.md
Comment (nicholassaylor) [reply worlds/dark_souls_3/docs/locations_en.md:194]: If the plan is to have it require Grave Key, then I would recommend that the logic require Grave Key, and have the drop be a trick that you could do outside of logic. 

If you want to keep the trick in logic, then I think that the note about it here is fine. Otherwise, I don't think it's necessary. I don't have a strong feeling either way, so you can mark this conversation as resolved when you've made a final decision.

File: worlds/dark_souls_3/docs/locations_en.md
Comment (nicholassaylor) [reply worlds/dark_souls_3/docs/locations_en.md:252]: If you think it's worth it, then I have no issue with it.

File: worlds/dark_souls_3/docs/en_Dark Souls III.md
Comment (nicholassaylor) [reply worlds/dark_souls_3/docs/en_Dark Souls III.md:144]: Ahh, okay I understand now after taking another look. I misunderstood the first bullet point. I would say that for the average user, the "without need progressive pickup lists" phrase is unnecessary. If you want to leave that in, I think it's fine. But you can resolve this conversation whenever you make a choice.

File: worlds/dark_souls_3/docs/setup_en.md
Comment (nicholassaylor) [reply worlds/dark_souls_3/docs/setup_en.md:14]: Then it might be a good idea to have the new client be a pre-release on Marech's repo until next AP release comes out.

File: worlds/dark_souls_3/docs/locations_en.md
Comment (nicholassaylor) [reply worlds/dark_souls_3/docs/locations_en.md:755]: I agree that it definitely helps with it being at the bottom. Another concern that I had was loading times for having such a large file, but after taking a look at the transfer size, the file ends up being surprisingly small, so I don't have any issues with it being here. I was more wondering if there was this table as part of the standalone.

File: worlds/dark_souls_3/__init__.py
Comment (qwint) [reply worlds/dark_souls_3/__init__.py:1305]: my point was more that any excluded locations under 'unrandomized' settings will return false for _is_location_available, and skip getting rules set, even if they are for a progression item
meaning that a location with an important progression item could be 'in logic' earlier than actually available to the player

File: worlds/dark_souls_3/__init__.py
Comment (qwint) [reply worlds/dark_souls_3/__init__.py:393]: ah, did not see those edit methods were returning new objects, will likely be fine then yea

File: worlds/dark_souls_3/__init__.py
Comment (qwint) [reply worlds/dark_souls_3/__init__.py:483]: I don't see Item having a reference to Multiworld, and a quick replacement of itempool in create_items() with self.itempool didn't fail any of the tests 

```
class DarkSouls3World(World):
    itempool: List[DarkSouls3Item]


    def __init__(self, multiworld: MultiWorld, player: int):
        self.itempool = []

    def set_rules(self) -> None:
        randomized_items = {
            item.name for item in self.itempool
        }
```
were my changes, plus the switches to self.itempool in create_items()

**Comment by qwint** [comment]
Just wanted to document my current state of review,
all my comments were fully addressed/clarified except:
I still dislike the Generate.py change being grouped in here as it has its own PR and from light testing this PR does not depend on it
My comment on saving the itempool for reference later instead of searching through the multiworld.itempool multiple times in the world - https://github.com/ArchipelagoMW/Archipelago/pull/3128#discussion_r1564391725 - not sure if the response got lost in the rest of the noise. I assume you were getting those errors because you initialized the list in the world itself, not in init/generate_early which is a very common mistake, but I still could be missing something

**Comment by nex3** [comment]
@PoryGone 

> Sneaking core changes through inside large World Content PRs isn't really a great solution.

The intention isn't to "sneak it in". This is just a PR that depends on another PR. Because this is large, I wanted to get the review started. I've filed #3167 for that, I'm sure it'll get in before this is ready.

@qwint

> I still dislike the Generate.py change being grouped in here as it has its own PR and from light testing this PR does not depend on it

This does depend on it—it's required to set the default `exclude_locations` option to a list of location groups. Those changes are being separately reviewed in https://github.com/ArchipelagoMW/Archipelago/pull/2403, but they're here as well so this PR can be tested in isolation.

> My comment on saving the itempool for reference later instead of searching through the multiworld.itempool multiple times in the world - https://github.com/ArchipelagoMW/Archipelago/pull/3128#discussion_r1564391725 - not sure if the response got lost in the rest of the noise. I assume you were getting those errors because you initialized the list in the world itself, not in init/generate_early which is a very common mistake, but I still could be missing something

Ah, initializing it in `create_items()` does work, thanks! Let me know if I missed any uses in 947d18c5e71172ba5515b8eed75e02b08a92ed68.

**Review by qwint — APPROVED** [review]
approving assuming this is properly blocked by #3167 #2403
I'm happy with the current updates

**Comment by Princesseuh** [comment]
Played a run using the latest beta, worked great! Just two things I encountered, apologies if already well known.

- Generating a yaml using the webhost in the fork didn't work, it generated `[]` somewhere it should've generated `{}`
- Path of the Dragon was randomized into the main shop in Firelink, I bought it, but Ididn't receive the emote.

Everything else worked perfectly fine and it was fun, awesome work!

File: worlds/dark_souls_3/docs/en_Dark Souls III.md
Comment (Marechal-L) [reply worlds/dark_souls_3/docs/en_Dark Souls III.md:198]: I would remove the second "have changed" category and put the "enable_*_locations" into the "removed" category while keeping the explanations on how to setup exclude_locations properly.
We would then be able to remove this block in a future PR when everyone has forgotten the 2.x.x

**Review by Marechal-L — COMMENTED** [review]
I confirm that generating a yaml from the webhost makes the random_enemy_preset to be assigned with [] instead of {} and generates an error.

Otherwise I strongly approve this version and I'll complete with a french translation of the docs

**Review by nex3 — COMMENTED** [review]
@Princesseuh

> * Generating a yaml using the webhost in the fork didn't work, it generated `[]` somewhere it should've generated `{}`

I think this is a bug in the option JSON generation. https://github.com/ArchipelagoMW/Archipelago/pull/3230 should fix it.

> * Path of the Dragon was randomized into the main shop in Firelink, I bought it, but I didn't receive the emote.

I'll take some time this week to investigate this issue.

**Review by qwint — APPROVED** [review]
saw my review got re-requested; skimmed the new changes since my last review, everything looks fine and my previous review stands
approving assuming this is properly blocked by https://github.com/ArchipelagoMW/Archipelago/pull/3167 https://github.com/ArchipelagoMW/Archipelago/pull/2403
I'm happy with it

**Review by nicholassaylor — APPROVED** [review]
The options changes look good!

The scope of the review is the same as https://github.com/ArchipelagoMW/Archipelago/pull/3128#pullrequestreview-1993290585 but I am giving my approval provided that everything that is blocking this is handled.

This PR should probably receive `waiting-on: other`

**Comment by ScipioWright** [comment]
> This PR should probably receive `waiting-on: other`

It already has `waiting-on: other`

**Comment by nicholassaylor** [comment]
> > This PR should probably receive `waiting-on: other`
> 
> It already has `waiting-on: other`

My bad, missed it from before

**Comment by nicholassaylor** [comment]
Since this PR no longer uses the flatten function, do the `Utils.py` changes need to be in this PR? It seems to make more sense to remove the `Utils.py` changes and since #2403 got merged, that would mean that this PR isn't blocked anymore.

**Review by Exempt-Medic — APPROVED** [review]
I reviewed all the changes and other than removing the changes to `Utils.py` and `AutoWorld.py` almost everything looks good to me. I have done thousands of generations on this branch over time and found a lot miscellaneous errors that have since been fixed. I did 500 more generation after merging this branch into `main` alongside the changes from [my own PRs](https://github.com/nex3/Archipelago/pulls). PRs 23 and 24 are fixes for things I noticed in my review.

I did **NOT** review `post_fill` but I believe it should be converted to `stage_post_fill` since it uses some expensive functions such as `get_spheres()`. I couldn't figure out a way to make that work, but it should be possible.

**Comment by qwint** [comment]
Skimmed the changes since I last approved and everything looked reasonable, and double checked my last blockers were resolved

File: worlds/dark_souls_3/__init__.py
Comment (NewSoupVi) [inline worlds/dark_souls_3/__init__.py:122]: Was local early items not an option here?

File: worlds/dark_souls_3/__init__.py
Comment (NewSoupVi) [inline worlds/dark_souls_3/__init__.py:495]: This is somewhat slow, is there a reason you can't wait to add self.local_itempool to self.multiworld.itempool until after this function runs, since those happen back to back in create_items anyway? Unless this function gets called somewhere else, but in that case, I'd potentially have a different problem with this

File: worlds/dark_souls_3/__init__.py
Comment (NewSoupVi) [inline worlds/dark_souls_3/__init__.py:539]: When checking for a location, region or entrance from an entrance access condition, unless the spot in question is part of the source region of the entrance, you need to register an indirect condition.

I think Medic is on the case already

Comment (NewSoupVi) [reply worlds/dark_souls_3/__init__.py:539]: You can use this:

https://github.com/NewSoupVi/Archipelago/tree/indirect_condition_checker

with this apworld & a DS3 yaml, to have it automatically detect missing indirect conditions

File: worlds/dark_souls_3/__init__.py
Comment (NewSoupVi) [inline worlds/dark_souls_3/__init__.py:1504]: This should probably check item rules too right?

Comment (NewSoupVi) [reply worlds/dark_souls_3/__init__.py:1504]: In fact, was location.can_fill(check_access=False) considered for a lot of these?

**Review by NewSoupVi — CHANGES_REQUESTED** [review]
I have a couple of points, and the item smoothing stuff is scary

code is overall pretty great tho

File: worlds/dark_souls_3/Items.py
Comment (NewSoupVi) [reply worlds/dark_souls_3/Items.py:227]: nvm

File: worlds/dark_souls_3/__init__.py
Comment (nex3) [reply worlds/dark_souls_3/__init__.py:539]: I'm a little confused here. All the location checks for these entrance rules are exclusively coming from the source regions (HWL is the source for Undead Settlement, RS is the source for Cathedral of the Deep, and so on). But from https://github.com/nex3/Archipelago/pull/27, it sounds like this isn't the whole story. Is there some documentation I can read to understand this better? Do we need to worry about the *transitive conditions* for all locations/regions/entrances that can affect a given entrance rule?

Part of the reason we use so many `self._can_get` calls is to abstract out the transitive complexity, since different locations can depend on fairly complex conditions to become accessible. It's also possible if we end up doing something like entrance randomizer that these conditions will end up extremely contingent on particular RNG outcomes. Should we think about refactoring this to avoid checking location accessibility at all, and instead produce events for things like "boss defeated" so we can avoid this issue entirely?

Comment (NewSoupVi) [reply worlds/dark_souls_3/__init__.py:538]: Idk how much you've been talking to Medic, they've been investigating the issue

I wasn't personally going to be able to confirm whether it's always the source region, but yes, you're right, if the entrance's source region and the location's parent region are the same, the indirect condition isn't necessary.

However, yes, you do need to also take care of transitive region requirements. The one case of this that actually tripped my indirect condition checker is this:

```
Dark Souls III, player 1: The parent region "Undead Settlement" of DarkSouls3Location "US: Soul of the Rotted Greatwood" could not be found in indirect_conditions for entrance "Go To Anor Londo" from "Irithyll of the Boreal Valley" to "Anor Londo". Suggested fix:
world.multiworld.register_indirect_condition(world.get_region("Undead Settlement"), world.get_entrance("Go To Anor Londo"))

This was caused by a recursive can_reach call.
The entrance access condition of "Go To Anor Londo" checked for
- DarkSouls3Location "IBV: Soul of Pontiff Sulyvahn", which checked for
- DarkSouls3Location "US: Soul of the Rotted Greatwood", which checked for
- the region in question, its parent region "Undead Settlement".
```

Now, it turns out that this is actually also fine, but in a non trivial way that my tool cannot account for. Medic explained to me that IBV can *only be reached through Undead Settlement*.

For an indirect condition to be necessary, two conditions have to be met:
- While inside an entrance access condition, a region is checked (This is transitive, so if an entrance checks a location, and that location checks another location, then that second location's region is also up for debate. I.e at some point before an entrance.can_reach has finished, somewhere down in the call stack, a region.can_reach is called)
- It's possible that the dependent region is reached "after" the source region of the entrance in a "random graph flood from menu region" kind of sense

Sorry for causing worry, I was not going to be able to confirm whether that was going to apply here or not, as I don't know anything about DS3.

If you want to read more about indirect conditions, the documentation is very lacking right now, but I have a PR open to remedy that somewhat: https://github.com/ArchipelagoMW/Archipelago/pull/3552
Although there I give a very "just make the indirect conditions" take, rather than the more nuanced "it might not always be necessary" version I just described to you (for public responsibility reasons)

Comment (NewSoupVi) [reply worlds/dark_souls_3/__init__.py:538]: The explanation for why indirect conditions are necessary is this.

Region sweep is basically a Breadth First Search of the region graph from the Menu region, checking entrances one by one and adding newly reached nodes (regions) to the queue until there is nothing more to check.

However, if entrance access conditions depend on regions, then it is possible for an entrance to be checked and determined to be intraversible because the region hasn't been reached, then after that the region is reached by the graph search.

To remedy this, we would have to recheck all entrances every time a new region is reached, until no new regions are reached.

Because most games do not check for region access inside of entrance access conditions, AP has decided to **eschew this rechecking** and just checks every entrance once. This gives a significant performance gain to AP as a whole, about 30%-50%.

However, because some games *did* start using things like `region.can_reach` inside entrance access conditions, we provided a way to **manually** define that a *specific* entrance needs to be rechecked during region sweep if a *specific* region is reached during it. This keeps almost all of the performance upsides, even a game making heavy use of indirect conditions is still way way faster than if it just blanket "rechecks all entrances until nothing new is found".

However, we do recognize it's a beginner's trap and some games are very complex with their access rules.
There is a PR that makes this behavior optional via a world class attribute: https://github.com/ArchipelagoMW/Archipelago/pull/3682

If you have any further questions, I'm happy to answer them here, in #ap-world-dev, or in DMs on Discord :)

Comment (NewSoupVi) [reply worlds/dark_souls_3/__init__.py:538]: Oh, to answer your last question:
Yes, if you have ER in the future the specific case in question with IBV and Undead Settlement could be a legitimate issue. The other cases are not gonna be an issue even with ER. If the source region of the entrance and the location's parent region are the same, it's 100% fine always.

Yes, events get around this issue, but they are **very slow**. If you can manage the indirect conditions, that'd be much much better.

Comment (NewSoupVi) [reply worlds/dark_souls_3/__init__.py:538]: Sorry, it's a big info dump, but I got the vibe that you might be interested in the "long version" haha

Comment (nex3) [reply worlds/dark_souls_3/__init__.py:530]: Thanks for the detailed background! It's very helpful to understand the underlying logic and trade-offs.

Is there a plan to pull the indirect condition checker into the main branch? I'm still a little nervous about missing conditions 😅

Comment (NewSoupVi) [reply worlds/dark_souls_3/__init__.py:530]: Probably not. It's a hugely hacked together monkey patched version of AP. I would just use it with an .apworld or something, that's what I did in this case

File: worlds/dark_souls_3/docs/locations_en.md
Comment (nicholassaylor) [inline worlds/dark_souls_3/docs/locations_en.md:46]:
```suggestion
* **Boss Rewards:** Boss drops. Does not include soul transfusions or shop
```
To be consistent with how it needs to be entered in a YAML

**Comment by NewSoupVi** [comment]
Exempt-Medic has made me aware of the issue with local shop items not being able to be sent to the server.

I think that's kind of bad, I think I would like you to investigate one of the following solutions:
1. Change all local shop items to events in `post_fill` by setting their address to `None` (Drawback: The multitracker now "spoils" which shop slots have local items)
2. Some sort of button in the client that "pseudo releases" all local shop item locations, after victory is achieved, or... something like that.

A room that has `!release` turned off should still be able to *theoretically* achieve 100% locations if possible.

(I'm aware of missable locations in DS3, which I'm also not entirely happy about, but at least you *could* make a new save for those if you *really* want 100%. With this shop issue, it seems that there is absolutely 0 ways to get some locations to be marked on the server)

**Comment by nex3** [comment]
Option 1 seems like the best compromise if you consider the status quo unacceptable, although I'm not entirely sure how to make that work while still passing the real item IDs to the offline randomizer.

Another option could be to have a "foreign shop items only" option. It couldn't be on by default because it intrinsically wouldn't work in a single-game randomizer, but it would fix multiworld cases. I really do want to get full support for proper shop item detection at some point, so hopefully we can find an acceptable compromise to get this landed in the meantime.

**Comment by nex3** [comment]
@NewSoupVi I've been mulling this over and over in my head and I finally cracked it. Local shop items are now properly marked as checked in the latest beta!

**Comment by NewSoupVi** [comment]
I think I will merge this soon, but under the "condition" that if https://github.com/ArchipelagoMW/Archipelago/pull/3739 gets in, `remaining_fill` is investigated for the post-fill smoothing algorithm.
(Which as I understand, Exempt-Medic already has a PR open for, to the DS3 AP repo, so that should be good.)

File: worlds/dark_souls_3/docs/setup_en.md
Comment (Mojken) [inline worlds/dark_souls_3/docs/setup_en.md:43]: Considering the last version worked on linux, a linux alternative should be provided for this batch file, and should be mentioned in these instructions

Comment (Mojken) [reply worlds/dark_souls_3/docs/setup_en.md:43]: Trying it out, it doesn't work because it looks for the path to the steam library in the windows registry... It's also been discontinued in favour of me3, but that in turn looks a bit too immature to start using already? Alternatively, it looks vaguely abandoned, judging by the commit history... Not sure what I think we should do, but relying on discontinued or abandoned software doesn't sound wise. An archipelago fork of ME2?

**Review by NewSoupVi — APPROVED** [review]
Fck it we ball - If there are issues with the item smoothing (the part about this that scares me the most) that come up in testing / during RC phase, my understanding is that it could just be turned off easily in an emergency bugfix PR.

100 test generations with a fully random DS3 yaml and 4 other yamls yielded no errors, so it seems fine.

**Pull request merged**
